packages/core/src/shapes/Shapes.ts

```diff
@@ -0,0 +1,169 @@
+import * as BBox from "engine/BBox";
+import { VarAD } from "types/ad";
+import { Value } from "types/value";
+import { Circle, makeCircle, sampleCircle } from "./Circle";
+import { Ellipse, makeEllipse, sampleEllipse } from "./Ellipse";
+import { Equation, makeEquation, sampleEquation } from "./Equation";
+import { Image, makeImage, sampleImage } from "./Image";
+import { Line, makeLine, sampleLine } from "./Line";
+import { makePath, Path, samplePath } from "./Path";
+import { makePolygon, Polygon, samplePolygon } from "./Polygon";
+import { makePolyline, Polyline, samplePolyline } from "./Polyline";
+import { makeRectangle, Rectangle, sampleRectangle } from "./Rectangle";
+import { Canvas, makeCanvas } from "./Samplers";
+import { Text, makeText, sampleText } from "./Text";
+
+//#region other shape types/globals
+
+// TODO: fix this type, it's too restrictive
+export interface Properties {
+  [k: string]: Value<VarAD>;
+}
+
+export type Shape =
+  | Circle
+  | Ellipse
+  | Equation
+  | Image
+  | Line
+  | Path
+  | Polygon
+  | Polyline
+  | Rectangle
+  | Text;
+
+export interface ShapeDef {
+  sampler: (canvas: Canvas) => Properties;
+  constr: (canvas: Canvas, properties: Properties) => Shape;
+
+  // TODO: maybe get rid of this?
+  propTags: () => { [prop: string]: Value<VarAD>["tag"] };
+
+  // TODO: make these methods
+  bbox: (properties: Properties) => BBox.BBox;
+  isLinelike: boolean; // TODO: use type predicate instead
+  isRectlike: boolean; // TODO: remove this
+}
+
+// hack to satisfy the typechecker
+export const ShapeDef = (shapedef: {
+  sampler: (canvas: Canvas) => unknown;
+  constr: (canvas: Canvas, properties: Properties) => Shape;
+  bbox: (properties: any) => BBox.BBox;
```

Comment:
Yes, because `constr` receives `Partial<IWhatever>` rather than just `IWhatever`. If you replace this `any` with `Properties` then you get type errors. It doesn't matter too much though, because we're soon going to replace this with a method on the shape anyway.